easybuild/tools/containers.py

```diff
@@ -0,0 +1,304 @@
+
+eb %(easyconfigs)s --robot --installpath=/app/ --prefix=/scratch --tmpdir=/scratch/tmp
+
+# exit from 'easybuild' user
+exit
+
+# cleanup
+rm -rf /scratch/tmp/* /scratch/build /scratch/sources /scratch/ebfiles_repo
+
+%%runscript
+eval "$@"
+
+%%environment
+source /etc/profile
+module use /app/modules/all
+module load %(mod_names)s
+
+%%labels
+
+"""
+
+
+_log = fancylogger.getLogger('tools.containers')  # pylint: disable=C0103
+
+
+def parse_container_base(base):
+    """Parse value passed to --container-base option."""
+    if base:
+        base_specs = base.split(':')
+        if len(base_specs) > 3 or len(base_specs) <= 1:
+            error_msg = '\n'.join([
+                "Invalid format for --container-base, must be one of the following:",
+                '',
+                "--container-base localimage:/path/to/image",
+                "--container-base shub:<image>:<tag>",
+                "--container-base docker:<image>:<tag>",
+            ])
+            raise EasyBuildError(error_msg)
+    else:
+        raise EasyBuildError("--container-base must be specified")
+
+    # first argument to --container-base is the Singularity bootstrap agent (localimage, shub, docker)
+    bootstrap_agent = base_specs[0]
+
+    # check bootstrap type value and ensure it is localimage, shub, docker
+    if bootstrap_agent not in SINGULARITY_BOOTSTRAP_TYPES:
+        known_bootstrap_agents = ', '.join(SINGULARITY_BOOTSTRAP_TYPES)
+        raise EasyBuildError("Bootstrap agent in container base spec must be one of: %s" % known_bootstrap_agents)
+
+    res = {'bootstrap_agent': bootstrap_agent}
+
+    for idx, base_spec in enumerate(base_specs[1:]):
+        res.update({'arg%d' % (idx + 1): base_spec})
+
+    return res
+
+
+def generate_singularity_recipe(easyconfigs, container_base):
+    """Main function to Singularity definition file and image."""
+
+    cont_path = container_path()
+
+    # check if --containerpath is valid path and a directory
+    if os.path.isdir(cont_path):
+        _log.info("Path for container recipes & images: %s", cont_path)
+    else:
+        raise EasyBuildError("Location for container recipes & images is a non-existing directory: %s" % cont_path)
+
+    base_specs = parse_container_base(container_base)
+
+    # extracting application name,version, version suffix, toolchain name, toolchain version from
+    # easyconfig class
+
+    bootstrap_agent = base_specs['bootstrap_agent']
+
+    base_image, base_image_tag = None, None
+
+    # with localimage it only takes 2 arguments. --container-base localimage:/path/to/image
+    # checking if path to image is valid and verify image extension is '.img' or '.simg'
+    if base_specs['bootstrap_agent'] == LOCALIMAGE:
+        base_image = base_specs['arg1']
+        if os.path.exists(base_image):
+            # get the extension of container image
+            image_ext = os.path.splitext(base_image)[1]
+            if image_ext == '.img' or image_ext == '.simg':
```

Comment:
@boegel

Should we also allow bootstrap from sandbox container. Sandbox container is just a directory so it is hard to check if a directory is a singularity container or just a regular directory. Something to think about for future updates to eb-singularity support.

Reply:
Well, the easy thing to do would be to just remove the extension check all together, and trust that the users knows what he/she is doing... Less code to worry about too ;)

Reply:
Let me ask with singularity if bootstraping with sandbox is a good idea, as for the logic I think having a check for local image is good. Ideally we need a singularity python api to identify something is a valid singularity container from shub, docker and sandbox container.